Add chart about what happens to entities when user/org is deleted (#42371)
sourcegraph/sourcegraph: doc/admin/user_data_deletion.md

- Bulk operations on changesets and batch changes created by the user.

> WARNING: If deleted and recreated a user will no longer have access to the configurations and customizations they may have made for their old account. The database will store information about their previous configuration, but a user associated with a new `id` in the `users` table in PostgreSQL will be created. For example, site admin privileges will have to be re-assigned if a once site admin user is deleted and recreated. 

## Entities deletion

> When a user or organization is deleted, here is what happens to their entities: 

Reviewer: Is there already documentation that describes the difference between soft and hard deletes?  We should probably specify which we are referring to when we say "is deleted".  It may be both and in that case could we explicitly say that?

Author: Soft and hard deletes are described above on this page! Will note this applies to both.

> When a user or organization is deleted, here is what happens to their entities: 

Author: ```suggestion
When a user or organization is deleted (for both "delete" or "delete forever"), here is what happens to their entities: 
```

> | Entity | Deleted when you delete the creator? | Details |

Reviewer: I'm not sure there's any value in the "Deleted when you delete the creator" column since we don't have any "No".  It seems like it's really either "Always" or "Sometimes". Are there any things that don't get deleted when a user is deleted?  
I'm not against keeping this as it is; just something I noticed.

Author: 2 reasons: 
1. I think it's useful to have a quick reference
2. In the future there _needs to be_ more things that are "no" – we've gotten feedback from users that objects should be re-assigned, not just deleted, and that's in the longer term plans for perms improvements and deletion improvements.

| ------ | ------------------------------------ | ------- |

> | Batch Change | Yes | If you delete the user or organization that owns the batch change, it gets deleted from our database and no longer appears in the UI. |

Reviewer: "it" here means "all batch changes created by this user"?

Author: ```suggestion
| Batch Change | Yes | If you delete the user or organization that owns the batch change, the batch change and all associated information gets deleted from our database and no longer appears in the UI. |
```

Edited! More details if you're curious https://docs.google.com/document/d/1xsgCRVIjgWz1vbSmHpv_3O2hvAFoxs_LKOCdGI9E_dg/edit#heading=h.z9pl9n59k717

| Code Insight chart | Sometimes | Private insights never shared with others are deleted. Insights shared to org-visible or global dashboards persist as long as the org and global dashboard continue to exist. |
| Code Insight dashboard | Sometimes | If the dashboard was private to the user it is deleted on the user deletion; if it was private to the org it is deleted on the org deletion. If it was a global dashboard, it will continue to exist. |
| Code Monitor | Yes | If the user that created the code monitor is deleted, their code monitors are deleted and will no longer trigger new actions. |
| Search Notebook | Yes | If you delete the user or organization that owns the notebook, it no longer appears in the UI. In the organization-owned case, it's still preserved in the database. |

> | Search Context | Yes | If you delete the user or organization that owns the context, it no longer appears in the UI. |

Reviewer: could we expand on "it no longer appears in the UI".  Does that also mean it's deleted from the database?  (asking because some customers ask if it's possible to go find the record in the db and run sql commands on their own to recover).
This comment applies to the other spots where "no longer appears in the UI" is mentioned.

Author: ```suggestion
| Search Context | Yes | If you delete the user or organization that owns the context, it no longer appears in the UI.  In cases where it was an organization notebook, it is preserved in the database. |
```

Sure, that was intentionally vague because this seems weirder to me, but this is what I have from Ben before he left. https://docs.google.com/spreadsheets/d/1yD6FQyfUOhtU-PyqdXQ27TdevgKJXhNWXRHVAo2bp0k/edit#gid=0

> | Settings file (extensions, experimental features, defaults) | Yes | If you delete the user organization, the associated settings file is deleted. |

Reviewer: did you mean "delete the user _or_ organization"?

Author: ```suggestion
| Settings file (extensions, experimental features, defaults) | Yes | If you delete the user or organization, the associated settings file is deleted. |
```
